src/NeoSharp.Core/Messaging/Handlers/BlockHeadersMessageHandler.cs

```diff
@@ -11,23 +13,51 @@ namespace NeoSharp.Core.Messaging.Handlers
     public class BlockHeadersMessageHandler : IMessageHandler<BlockHeadersMessage>
     {
         private readonly IBlockchain _blockchain;
+        private readonly IBinarySerializer _serializier;
+        private readonly ICrypto _crypto;
         private readonly ILogger<BlockHeadersMessageHandler> _logger;
 
-        public BlockHeadersMessageHandler(IBlockchain blockchain, ILogger<BlockHeadersMessageHandler> logger)
+        public BlockHeadersMessageHandler(IBlockchain blockchain, ILogger<BlockHeadersMessageHandler> logger, IBinarySerializer serializier, ICrypto crypto)
         {
             _blockchain = blockchain ?? throw new ArgumentNullException(nameof(blockchain));
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _serializier = serializier ?? throw new ArgumentNullException(nameof(serializier));
+            _crypto = crypto ?? throw new ArgumentNullException(nameof(crypto));
         }
 
         public async Task Handle(BlockHeadersMessage message, IPeer sender)
         {
-            _blockchain.AddBlockHeaders(message.Payload.Headers.Select(u => u.Header));
+            // Why we need to store BlockHeaderBase? don't have TX hashes, is better ask for blocks and store when we receive it
```

Comment:
Ok, I looked into NEO project to understand why this works like this:
[https://github.com/neo-project/neo/blob/9cf0d94f0979ea579e6b9d660359e0c49625532b/neo/Network/RemoteNode.cs#L227](https://github.com/neo-project/neo/blob/9cf0d94f0979ea579e6b9d660359e0c49625532b/neo/Network/RemoteNode.cs#L227)

## How we sync block headers

**TL;DR**: VerAckMessageHandler kicks of the cycle of getting block headers and BlockHeadersMessageHandler continues it until gets them all.

GetBlockHeadersMessageHandler/BlockHeadersMessageHandler work in conjunction with VerAckMessageHandler. If the current peer has fewer block headers than the remote one it requests missing block headers sending GetBlockHeadersMessage in VerAckMessageHandler. We expect to get the result in BlockHeadersMessage and to handle in BlockHeadersMessageHandler. BUT, GetBlockHeadersMessageHandler has limitation on the number block headers in response (MaxBlockHeadersCountToReturn = 2000). I suspect it is introduced as the pagination mechanism. This is the reason why we have the duplicate condition in VerAckMessageHandler and BlockHeadersMessageHandler.

## Why we don't care about transaction hashes in block headers

**TL;DR**:  transaction hashes are not critical for blockchain sync.

We have two separate messages for blocks and block headers. Consequently, they are synchronized differently and have different applications. The goal for block headers to sync as fast as possible to get the latest block hash to process other blocks based on it. So on start, the neo client is very busy to get block headers in sync. As opposed to that process, the goal for block sync is to get full fledge blockchain snapshot including transaction information. Needless to say, the process is heavy and is paused when we need to handle new block, transactions, consensus. The block sync operates just a block hash range for fetching. As you see this approach does not rely on transaction hashes. Maybe, we can use them from block headers for different purposes but right now I am not aware of these scenarios.

## Where the block sync is

**TL;DR**:  Not exists for now but we put this in PeerMessageListener

We are going to implement block sync as it was before. Please check neo project for reference
[https://github.com/neo-project/neo/blob/master/neo/Network/RemoteNode.cs#L456](https://github.com/neo-project/neo/blob/master/neo/Network/RemoteNode.cs#L456).
Our place is PeerMessageListener (in while loop). Later it will be optimized not to delay message processing.

**This change will break block header sync logic and every new peer will be overflooded with getting blockchain snapshot until it is able to do any useful workload. So please don't.**

Reply:
So i will made some changes, according to your comment

Reply:
Now i understand the dummy values :P

Reply:
@osmirnov please take a look again, i think that we can change Block like this

```
block
{
 BlockHeader Header{get;set;}
 Transaction[] Tx{get;set;}
}
```

Something like that ..